Adjust ncpus appropriately

milljm opened the pull request:
After applying the math to adjust node count for min_parallel, set ncpus
to what the end user supplied.

Closes #17204

Commit:
Adjust ncpus appropriately

After applying the math to adjust node count for min_parallel, set ncpus
to what the end user supplied.

Closes #17204

permcody commented on python/TestHarness/schedulers/RunPBS.py:
```diff
@@ -84,6 +84,7 @@ def _augmentTemplate(self, job):
         # Compute node requirement
         if self.options.pbs_node_cpus:
             nodes = template['mpi_procs']/self.options.pbs_node_cpus
+            template['mpi_procs'] = self.options.pbs_node_cpus
```
This is fine, but more work may be needed in the future to support more control of partial node allocations due to memory constraints. 

See the good 'ole cluster launcher script. Users can specify either just the number of "total mpi procs", or they can specify a more complete select specification with "mpi procs" being a per node number. We may need to think about this a bit more for the TestHarness now that we are moving to test larger and larger inputs since some of these may not work when using all the processors on a node.

bwspenc replied:
This doesn't seem quite right. The way I read this, if we specified `--pbs-node-cpus`, no matter how many cpus we asked for in a test, we're always going to get `pbs_node_cpus` times the number of nodes. If we had a small test that needed 4 cpus, we'd end up asking for 48 (assuming that's what we had specified in `--pbs-node-cpus`). Am I understanding that right?

milljm replied:
Let me just copy what the actual qsub file looks like, according to a multi-test spec file with different attributes for each test.

Given the following Test Spec file:

```conf
[Tests]
  [./always_ok]
    type = RunApp
    input = good.i
    min_parallel = 72
    max_time = 300
  [../]
  [./always_ok2]
    type = RunApp
    input = good.i
    min_parallel = 36
    max_time = 18000
  [../]
[]
```
Will produce:
```
#!/bin/bash
#PBS -N always_ok
#PBS -l select=3:ncpus=72
#PBS -l walltime=05:05:00
#PBS -P moose

#PBS -j oe
#PBS -o /Users/milljm/projects/moose/test/tests/test_harness/qsub.output
#PBS -l place=free

JOB_NUM=${PBS_JOBID%\.*}

export MV2_ENABLE_AFFINITY=0

cd /Users/milljm/projects/moose/test
/Users/milljm/projects/moose/test/run_tests -j 72 --pbs-node-cpus 32 --spec-file /Users/milljm/projects/moose/test/tests/test_harness/always_ok -o /Users/milljm/projects/moose/test/tests/test_harness
```

Lets dissect this a bit.

walltime is the culmination of all max_times. select= is the node count using simple whole integer math provided by the end user as --pbs-node-cpus, divided by the largest min_parallel in the group. ncpus is _now_ what the user specified it is, instead of the largest min_parallel (that is a bug and is wrong, as falcon in the case it was reported from does not have 72 cores per node).

Jumping down to ./run_tests -j 72. This is obvious. You have a test which requires 72 cores, so -j 72 is a minimum. Because there is more than one test, one will be waiting for the other to complete before it can start (not enough slots with -j 72. We would have to make this -j 96 for both to launch simultaneously).

I think folks are forgetting that the TestHarness cannot launch individual PBS job submissions as single tests from multi-test spec files. It can only launch the entire group as a single PBS job submission. And allow the TestHarness to do its job, due to test having `pre_req` capabilities, and facilitate the job allocation while inside a PBS job submission.

milljm replied:
Maybe we can add some additions that only affect spec files with larger min_parallel counts, and leave the smaller ones alone?

milljm replied:
@bwspenc how about this?

bwspenc replied:
I didn't realize that they get submitted as a group. When I run `./run_tests`, does that generate a single group of tests that gets run within a single PBS job, or multiple groups?  This might be easier to discuss over a call.

permcody reviewed (dismissed):
I have a script that you

bwspenc commented:
Just curious -- is there any way we can query pbs to find out how many cpus/node there are on the queue we're submitting to so we don't have to specify `--pbs-node-cpus`?  Based on a small amount of searching, I'm guessing the answer is no, but just checking.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/695477/) on e3f8f5e wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/17205/site)

This comment will be updated on new commits.

milljm commented:
The TestHarness has to cobble together all the tests provided in the spec file to generate a generic select statement that satisfies all tests contained therein. At best, it is only optimized for the one test that triggered a floor/ceiling change.

Please be aware, that all the TestHarness is doing, is launching the TestHarness _again_ while on a node and running that single test spec file:

```bash
./run_tests test-spec-file -j <whatever the ceiling is>
```
This means if you have 2 or more tests, one or more will wait while the single test which produced the maximum core requirement (the ceiling) completes.

I don't know how the ClusterLauncher works exactly, but looking over all the required params, it looks very specific to a 'ClusterLauncher-only' spec file. Making the ClusterLauncher very good at launching a single test with an optimal select statement for that single test.

milljm commented:
> Just curious -- is there any way we can query pbs to find out how many cpus/node there are on the queue we're submitting to so we don't have to specify `--pbs-node-cpus`? Based on a small amount of searching, I'm guessing the answer is no, but just checking.

There is. Looks like some regex/parsing would be involved.

```
pbsnodes -aSj
```

Commit:
Leave smaller jobs alone

Only create 'Chunks' when required. That is, if the job requires less than the
number of cores being specified with --pbs-node-cpus, do not alter the select
statement (select=1:ncpus=min_parallel).

Refs #17204

bwspenc commented:
I'm happy with the revised version -- that addresses my concern.
